### PR TITLE
Phase out `log_error` macro

### LINKED ISSUE #553: Remove instances of `log_error!`
Logging is a form of error handling. By logging the error in the handler, we prematurely perform a form of error handling, thus making it impossible to chose a different form of error handling at other call sites.

Goal: Remove all usages of `log_error!` and log them on the call-site of the message sending. In a later stage, we can then change the error handling to something different.

### PR DESCRIPTION
**Pull request by luckysori**
Fixes #553.

### CODE REVIEW

File: daemon/src/send_future_ext.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    async fn warn(self);
+}
+
+#[async_trait]
+impl<E: Error + Send, A: xtra::Actor, M: xtra::Message<Result = Result<(), E>>> Warn
+    for xtra::address::SendFuture<A, M>
+{
+    async fn warn(self) {
+        let msg = match self.await {
+            Err(e) => format!("{:#}", e),
```
Comment (thomaseizinger) [inline daemon/src/send_future_ext.rs]: I think we should still return`Disconnected` here.

File: daemon/src/send_future_ext.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+use async_trait::async_trait;
+use std::error::Error;
+
+#[async_trait]
+pub trait Warn {
+    async fn warn(self);
```
Comment (thomaseizinger) [inline daemon/src/send_future_ext.rs]: Perhaps pass in a `&str` so we  can add a message to the error?

File: daemon/src/lib.rs [hunk 3]
```diff
@@ -47,6 +47,7 @@ pub mod projection;
 pub mod rollover_taker;
 pub mod routes;
 pub mod seed;
+pub mod send_future_ext;
```
Comment (thomaseizinger) [inline daemon/src/lib.rs]:
```suggestion
pub mod xtra_ext;
```

Probably go after the crate.

File: daemon/src/send_future_ext.rs (hunk 2 above)
Comment (thomaseizinger) [reply daemon/src/send_future_ext.rs]:
```suggestion
    async fn log_failure(self);
```

A bit longer but I think easier to understand.

**Review by thomaseizinger — COMMENTED** [review]
I think that looks quite good already!

Some suggestions.

File: daemon/src/send_future_ext.rs (hunk 1 above)
Comment (luckysori) [reply daemon/src/send_future_ext.rs]: I don't think we've been making a clear distinction in terms of handling actor disconnects and handler failures. Doing this would suggest to me that we should almost always bubble up actor disconnects. Is that what you see?

Otherwise, wouldn't this change just make this trait method less convenient?

File: daemon/src/send_future_ext.rs (hunk 2 above)
Comment (luckysori) [reply daemon/src/send_future_ext.rs]: > Perhaps pass in a `&str` so we can add a message to the error?

What does the final log string look like in your mind. I'm not convinced we need this, but maybe it could be an `Option`.

File: daemon/src/send_future_ext.rs (hunk 1 above)
Comment (thomaseizinger) [reply daemon/src/send_future_ext.rs]: We should be making a distinction as in most cases, an actor being disconnected is a bug I'd say? Now the question is, do we want to handle that in one place or explicitely at every call site?

The only example I can think of where it might not be a bug is if we tie the lifetime of a network connection to an actor.

File: daemon/src/send_future_ext.rs (hunk 2 above)
Comment (thomaseizinger) [reply daemon/src/send_future_ext.rs]: I see it as an equivalent to `anyhow::Context`.

Very likely, a fallible handler has more than one `?` in it so the overall context of what bigger operation failed is lost.

What we could do is introduce a trait similar to `ActorName` for messages and include a generic message here that says `"Message handler for {} failed: {:#}"`.

Probably can even unify that to a single `Name` trait or - dare I say it - use `Display` 😁

File: daemon/src/oracle.rs [hunk 4]
```diff
@@ -195,7 +196,10 @@ impl Actor {
     ) -> Result<()> {
         tracing::info!("Fetched new attestation for {}", id);
 
-        let _ = self.attestation_channel.send(attestation).await;
+        self.attestation_channel
+            .send(attestation)
+            .log_failure()
+            .await?;
```
Comment (thomaseizinger) [inline daemon/src/oracle.rs]: To preserve semantics, we should do `let _ = ` here.

File: daemon/src/fan_out.rs [hunk 5]
```diff
@@ -1,32 +1,92 @@
...
 }
 
-impl<M> xtra::Actor for Actor<M> where M: xtra::Message<Result = ()> {}
+impl<M> xtra::Actor for Actor<M> where M: xtra::Message<Result = Result<()>> {}
 
 #[async_trait::async_trait]
 impl<M> xtra::Handler<M> for Actor<M>
 where
-    M: xtra::Message<Result = ()> + Clone + Sync + 'static,
+    M: xtra::Message<Result = Result<()>> + Clone + Sync + 'static,
```
Comment (thomaseizinger) [inline daemon/src/fan_out.rs]: Is this change necessary?

File: daemon/src/fan_out.rs [hunk 6]
```diff
@@ -1,32 +1,92 @@
...
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn errors_display() -> Result<()> {
```
Comment (thomaseizinger) [inline daemon/src/fan_out.rs]: Can we assert the actual display impl here?

File: daemon/src/send_future_ext.rs (hunk 1 above)
Comment (luckysori) [reply daemon/src/send_future_ext.rs]: I agree that it's always unexpected when the receiving actor is down. But I think we have been primarily logging these errors, so I suspect if we return the `Disconnected` error we're just gonna log at every call site.

If you want the flexibility of being able to handle the `Disconnected` error, could we introduce another trait which warns about the disconnect. I'm just worried about ergonomics and code readability if we just match against the error everywhere.

File: daemon/src/fan_out.rs (hunk 6 above)
Comment (luckysori) [reply daemon/src/fan_out.rs]: Sorry, I was just testing things for myself. Do you think it's worth it to keep this test?

File: daemon/src/oracle.rs (hunk 4 above)
Comment (luckysori) [reply daemon/src/oracle.rs]: Do we want to preserve the semantics though? We weren't taking advantage of returning a `Result` before. I can put in a separate commit if you prefer.

Comment (thomaseizinger) [reply daemon/src/oracle.rs]: I think in this first transition yes. We can then re-evaluate, where we fail for what reason.

File: daemon/src/fan_out.rs (hunk 6 above)
Comment (thomaseizinger) [reply daemon/src/fan_out.rs]: Well, I am not convinced we need this entire construct? Do we need to change `fan_out` to return a result? Note that I think we can also just drop the usage of `fan_out` if it is too painful.

File: daemon/src/fan_out.rs (hunk 5 above)
Comment (luckysori) [reply daemon/src/fan_out.rs]: I don't think it's strictly necessary, but I think it goes with the theme of phasing out `log_error`:

1. We want to, in general, shift from logging at the handler to returning errors to the sender.
2. The `fan_out::Actor` was preventing us from passing errors through back to the sender.

I think it would be better if we could fan out messages of both types here i.e. `()` and `Result<()>`. Is there any way we could do that?

Comment (thomaseizinger) [reply daemon/src/fan_out.rs]: > We want to, in general, shift from logging at the handler to returning errors to the sender.

Yes but in this case, `fan_out` was explicitly designed to only work with messages that never fail to avoid exactly the decision making of "what do we do if one of the messages fails?"

So unless this is a change motivated by another change that is more related to the `log_error` macro, I'd leave that in place for now.

File: daemon/src/send_future_ext.rs (hunk 2 above)
Comment (luckysori) [reply daemon/src/send_future_ext.rs]: I prefer `Name` here since we probably only want to log the name of the type. If not I would use `Debug`.

File: daemon/src/oracle.rs (hunk 4 above)
Comment (luckysori) [reply daemon/src/oracle.rs]: I've tried to follow your advice and preserve semantics all across the board. But I may have slipped, so do keep an eye out for mistakes like this one in your next review.

**Comment by luckysori** [comment]
I'm not sure if I will want to keep all these commits, but they should help at review time, @thomaseizinger.

File: daemon/src/oracle.rs [hunk 7]
```diff
@@ -308,6 +309,12 @@ impl xtra::Message for Attestation {
     type Result = Result<()>;
 }
 
+impl xtra_ext::Name for Attestation {
+    fn name() -> String {
+        "Attestation".to_string()
+    }
+}
```
Comment (thomaseizinger) [inline daemon/src/oracle.rs]: This wants to be a custom derive :D

File: daemon/src/oracle.rs [hunk 8]
```diff
@@ -257,8 +259,13 @@ pub struct NoAnnouncement(pub BitMexPriceEventId);
 
 #[async_trait]
 impl xtra::Handler<NewAttestationFetched> for Actor {
```
Comment (thomaseizinger) [inline daemon/src/oracle.rs:260]: Perhaps make a note somewhere that this can use `xtra-productivity` now.

File: daemon/src/monitor.rs [hunk 9]
```diff
@@ -666,8 +665,8 @@ impl<C> xtra::Handler<Sync> for Actor<C>
 where
     C: bdk::electrum_client::ElectrumApi + Send + 'static,
 {
-    async fn handle(&mut self, _: Sync, _ctx: &mut xtra::Context<Self>) {
-        log_error!(self.sync());
+    async fn handle(&mut self, _: Sync, _ctx: &mut xtra::Context<Self>) -> Result<()> {
```
Comment (thomaseizinger) [inline daemon/src/monitor.rs]: Does this actually compile by itself? We are sending this `Sync` message using notify_interval I believe (which would be nice if we would start this from the `started()` lifecycle hook. If this compiles, then I think we might kill the sync if we return a `Result` because then this loop might stop.

File: daemon/src/maker_inc_connections.rs [hunk 10]
```diff
@@ -246,10 +247,10 @@ impl Actor {
         tasks.add(write_fut);
         self.connection_tasks.insert(taker_id, tasks);
 
-        let _ = self
-            .taker_connected_channel
+        self.taker_connected_channel
             .send(maker_cfd::TakerConnected { id: taker_id })
-            .await;
+            .log_failure("Failed to report new taker connection")
+            .await?;
```
Comment (thomaseizinger) [inline daemon/src/maker_inc_connections.rs]: Not semantically equivalent.

File: daemon/src/maker_inc_connections.rs [hunk 11]
```diff
@@ -136,10 +136,15 @@ impl Actor {
...
-            let _ = self
+            if self
                 .taker_disconnected_channel
                 .send(maker_cfd::TakerDisconnected { id: *taker_id })
-                .await;
+                .log_failure("Failed to inform about taker disconnect")
+                .await
+                .is_err()
+            {
+                tracing::warn!("TakerDisconnected receiver actor is disconnected");
```
Comment (thomaseizinger) [inline daemon/src/maker_inc_connections.rs]: I am not sure we really need those? Bloats things quite a bit.

File: daemon/src/rollover_maker.rs [hunk 12]
```diff
@@ -140,7 +140,7 @@ impl Actor {
             order_id: self.cfd.id(),
             dlc,
         };
-        self.maker_cfd_actor.send(msg).await?;
+        self.maker_cfd_actor.send(msg).await??;
```
Comment (thomaseizinger) [inline daemon/src/rollover_maker.rs]: This should use `.log_error`?

File: daemon/tests/harness/mod.rs [hunk 13]
```diff
@@ -367,9 +367,13 @@ impl Taker {
 #[macro_export]
 macro_rules! deliver_event {
     ($maker:expr, $taker:expr, $event:expr) => {
-        tracing::debug!("Delivering event: {:?}", $event);
-        $taker.system.cfd_actor_addr.send($event).await.unwrap();
-        $maker.system.cfd_actor_addr.send($event).await.unwrap();
+        #[allow(unused_must_use)]
```
Comment (thomaseizinger) [inline daemon/tests/harness/mod.rs:370]: Why? Shouldn't we `unwrap` here?

**Review by thomaseizinger — APPROVED** [review]
Nice work! Definitely some follow-up things that we can do here.

I think most of this is sane :)

**Comment by thomaseizinger** [comment]
> I'm not sure if I will want to keep all these commits, but they should help at review time, @thomaseizinger.

Sorry for the late review, I got carried away coding half way through!
The individual commits were definitely helpful, but I think we should squash for the merge.

File: daemon/src/oracle.rs (hunk 7 above)
Comment (luckysori) [reply daemon/src/oracle.rs]: For simplicity let's just settle on using `Display`. We can always tweak the `Display` implementations if we want more out of them.

File: daemon/src/oracle.rs (hunk 8 above)
Comment (luckysori) [reply daemon/src/oracle.rs:260]: I've opened #905 to keep track of this in general.

File: daemon/src/monitor.rs (hunk 9 above)
Comment (luckysori) [reply daemon/src/monitor.rs]: Okay, I think I get you. The compiler doesn't complain, but I'll revert this change and log in the handler like we used to.

I think it wouldn't actually kill the notification loop since it ignores the variant of the inner result:

```rust
async move {
    Delay::new(duration).await;
    while addr.do_send(constructor()).is_ok() {
        Delay::new(duration).await
    }
};
```

But we would lose the logging of the error nonetheless.

File: daemon/src/maker_inc_connections.rs (hunk 11 above)
Comment (luckysori) [reply daemon/src/maker_inc_connections.rs]: I agree about the bloat, which is why my first thought was to include handling this in the `log_failure` trait. So you would just ignore these errors?

Comment (thomaseizinger) [reply daemon/src/maker_inc_connections.rs]: At the moment we are ignoring them right? So I don't think we are degrading anything if we continue to ignore them!

It is hard to introduce changes like the current one if you are at the same time trying to change behaviour too much.

Reason I would ignore them in most cases is because they can never happen as by default, actors don't shut down in xtra. It is an explicit action.

File: daemon/tests/harness/mod.rs (hunk 13 above)
Comment (luckysori) [reply daemon/tests/harness/mod.rs:370]: Unfortunately we cannot `unwrap` twice unconditionally, as sometimes sending only returns `Result<(), Disconnected>`.

File: daemon/src/maker_inc_connections.rs (hunk 11 above)
Comment (luckysori) [reply daemon/src/maker_inc_connections.rs]: > It is hard to introduce changes like the current one if you are at the same time trying to change behaviour too much.

I agree on principal, but logging instead of silently dropping is never going to have unintended side-effects.

> Reason I would ignore them in most cases is because they can never happen as by default, actors don't shut down in xtra. It is an explicit action.

What if we explicitly shut down the receiving end and still want to send messages there? In this case I agree that we can ignore the error, but it seems wrong as a general rule.

Comment (luckysori) [reply daemon/src/maker_inc_connections.rs]: It the argument is that it can never happen, shouldn't we be `expect`ing?

**Comment by luckysori** [comment]
@thomaseizinger: One final review (hopefully) and then I'll squash everything and complete the rebase.

File: daemon/src/oracle.rs (hunk 7 above)
Comment (luckysori) [reply daemon/src/oracle.rs]: I think `std::any::type_name` may be all we need for logging and way more concise.

File: daemon/src/xtra_ext.rs [hunk 14]
```diff
@@ -33,22 +33,18 @@ where
...
-    M: xtra::Message<Result = anyhow::Result<()>> + Name,
+    M: xtra::Message<Result = anyhow::Result<()>>,
 {
     async fn log_failure(self, context: &str) -> Result<(), Disconnected> {
         if let Err(e) = self.await? {
             tracing::warn!(
                 "{}: Message handler for message {} failed: {:#}",
                 context,
-                M::name(),
+                std::any::type_name::<M>(),
```
Comment (thomaseizinger) [inline daemon/src/xtra_ext.rs:43]: image

**Review by thomaseizinger — APPROVED** [review]

File: daemon/src/maker_inc_connections.rs [hunk 15]
```diff
@@ -354,32 +357,26 @@ impl Actor {
...
+                let _ = self
                     .settlement_actors
                     .send(&order_id, collab_settlement_maker::Initiated { sig_taker })
-                    .await
-                    .is_err()
-                {
-                    tracing::warn!(%order_id, "No active settlement");
-                }
+                    .await;
             }
```
Comment (thomaseizinger) [inline daemon/src/maker_inc_connections.rs]: Sorry to bring this up only now but this change is kind of unintended. The (ephemeral) actors being disconnected here is very much possible which is why I would like to keep these messages.

File: daemon/tests/harness/mod.rs (hunk 13 above)
Comment (thomaseizinger) [reply daemon/tests/harness/mod.rs:370]: Ah that is a shame. @klochowicz are you okay with this for now? You introduced the macro but unfortunately, it is no longer as generically applicable. We might need two macros (`deliver_fallible_event` and `deliver_event`) or create some other abstraction on the actual `Maker` and `Taker` structs.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
One comment, otherwise good to go!

File: daemon/src/maker_inc_connections.rs (hunk 15 above)
Comment (luckysori) [reply daemon/src/maker_inc_connections.rs]: Ahh, my bad. I thought I didn't mess this up.

**Review by thomaseizinger — APPROVED** [review]
